pkg/gce-cloud-provider/compute/fake-gce.go

```diff
@@ -422,6 +421,42 @@ func (cloud *FakeCloudProvider) getGlobalSnapshotURI(snapshotName string) string
 		snapshotName)
 }
 
+type FakeBlockingCloudProvider struct {
+	*FakeCloudProvider
+	snapshotToExec chan SnapshotSourceAndTarget
```

Comment:
Okay, I understand what your test is doing now.
1. make concurrent API calls
2. assert that conflicting operations do not happen concurrently

The fact that you return an aborted error makes the test a bit easier to write. If you assert an aborted error, you know that (a) the race condition is prevented, and (b) the test actually covered the race condition.

Synchronizing a test like this with channels is the right approach IMO, but it can be structured in a way that is easier to read and removes the tight poll loop.

My number one piece of advice is **prefer callbacks instead of channels in APIs**. (There's a really good gophercon talk about this, which I can't find at the moment.) An API that uses channels can have many ways to mis-use it. Should API users read, write, or close the channels? Do callers have control over the buffer size? Buffered vs unbuffered can make a semantic difference. It's difficult to picture how `FakeBlockingCloudProvider` would be used by a different unit test.

An API that uses channels to communicate can be transformed into a callback-based API. These are much more test friendly because callbacks can be wrapped in funcs that can interact with the test environment before and after the actual callback runs.

In this specific example, you could have a func value in the struct called `beforeSnapshot` which is called before you delegate into `FakeCloudProvider.CreateSnapshot`. That func can come from your test, where it is able to interact with local state, such as synchronization channels.

Reply:
The thing with call backs is that while I guess it makes it more generic, it just moves the same logic that was in the BlockingCloudProvider struct into the test. I don't think it removes any of the complexity that's the issue,.

Reply:
Splitting test logic across two files *is* complexity. As we discussed, I don't think your wrapper around the fake cloud provider needs to be in a separate file.